server/src/main/java/org/apache/druid/client/CachingClusteredClient.java

```diff
@@ -618,6 +619,7 @@ private void addSequencesFromServer(
         final SortedMap<DruidServer, List<SegmentDescriptor>> segmentsByServer
     )
     {
+      responseContext.put(Key.REMAINING_RESPONSES_FROM_QUERY_NODES, segmentsByServer.size());
```

Comment:
I think the response contexts are shared between subqueries, but it won't be a problem because subqueries are not executed in parallel for now. Same for `UnionDataSource`. However, I agree that it should be per subquery for the future. I made needed changes to set subqueryId for union and store number of servers remaining per subquery or query.

Reply:
Oh, I actually found the response context is not shared between subqueries when inlining them. It is shared in UnionQueryRunner though. I think it makes sense to not share the context when inlining subqueries.